### ISSUE TITLE
ORM Schema generator needs a way of specifying initialising records for a new table

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-framework#8908]
## Problem

When the schema migrator in `dev/build` adds new tables to an existing site, those tables will sometimes need to have data added to them to ensure that data in the other tables retains integrity. This has shown up in some upgrade bugs and some edge case ORM bugs that have been hanging around a while. 

## Use-cases

 * When a Page table is created for the first time, it should be initialised with empty records for the relevant SiteTree records
 * When Version is added Obj_Live and Obj_Versions should be initialised with data to have a single, published version for each Obj. (this particularly affects SS3 -> SS4 upgrades in File)

## Possible approaches

...tbc..

## Bugs it will fix

 * https://github.com/silverstripe/silverstripe-framework/issues/4700
 * https://github.com/silverstripe/silverstripe-framework/issues/7277
 * https://github.com/silverstripe/silverstripe-upgrader/issues/163

**Comment by sminnee** [comment]
@dnsl48 do you have any views on how to approach this?

**Comment by sminnee** [comment]
To be honest having written out the options, I'm kind of inclined to go with requireDefaultRecords, as the disadvantages aren't huge, and will be a bit more of a JFDI solution than designing and introducing new APIs.

**Comment by chillu** [comment]
If we put this on the critical path for "solve file migration once and for all in 4.4.0", then we have just over a week to get this in. So I'm strongly in favour of not overcooking this.

**Comment by sminnee** [comment]
Notes about the the requireDefaultRecords approach:

 * DataObject::requireDefaultRecords will be called on each object, after all the schema updates have been made, so you can expect all the queries to exist
 * Versioned::requireDefaultRecords is called by the extend method at the end of DataObject::requireDefaultRecords
 * Both of these methods will be called on each subclass in turn

**Comment by dnsl48** [comment]
Yes, agree, `requireDefaultRecords` feels natural and its description does already look like it supposed to solve that exact issue.

> Con: Have to have special checks to determine "has this table just been created or not?

On the other hand developers have the full control over it, which is a good default. It won't limit the ones who might need something trickier. The other way feels riskier as we don't have declarative migrations which users would be able to revert and reapply.

**Comment by sminnee** [comment]
Great, sounds like we're in agreement.

**Comment by tractorcow** [comment]
I feel this problem could be split into a few sub-tasks:
 - Ensure that `$defaults` config is properly emitted to db schema construction. I.e. if I have a `defaults` value of `1` for an int field, I'd like that to be created as `INT(11) default 1`. At the moment, defaults only generates the default in memory.
- Ensure that missing rows are added when appropriate for subtyped classes. This happens not only at schema generation (as you'd mentioned above) but also when changing classes.

I would probably look at solution 1 `DB::require_table()` if I had to choose, but I'd much rather a pluggable "build pipeline" we've needed for years. We keep patching stuff into requireDefaultRecords() that shouldn't be there, when what we really need is user customisable build steps. (e.g. setup, before build , after build, cleanup) that you can plug with yml.

**Comment by dnsl48** [comment]
> user customisable build steps. (e.g. setup, before build , after build, cleanup)
I _think_ in that case we would still have to manually check for "has this table just been created or not?".
In that case the whole implementation of `requireDefaultRecords` would simply move into `after build` step?
Otherwise, there could be some tricky edge cases, e.g.
 - developer deployed a bug so that a table got created, but not populated, so the next build won't trigger the "just after table generation" step.
 - one table defaults reference another table defaults, but the former gets populated first by the dev/build, so user needs to trigger dev/build the second time, however both tables are already created

**Comment by sminnee** [comment]
I started doing a PoC implementation of this a few days ago. One of the stumbling blocks I came across is that our query builder doesn't currently have any support for `INSERT INTO ... SELECT` statements.

Most of the SQLInsert class is designed to support rows of value literals. Probably a separate SQLInsertSelect class is better than shoehorning the functionality into SQLInsert. DBQueryBuilder will need to be modified to handle this, probably by adding a separate `buildInsertSelectQuery` method.

**Comment by sminnee** [comment]
I'd note that this would likely fix #7072
